flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerTest.java

```diff
@@ -0,0 +1,415 @@
+			firstDispatcherLeaderProcess,
+			secondDispatcherLeaderProcess);
+
+		try (final DefaultDispatcherRunner dispatcherRunner = createDispatcherRunner()) {
+			testingLeaderElectionService.isLeader(firstLeaderSessionId);
+
+			final CompletableFuture<DispatcherGateway> firstDispatcherGatewayFuture = dispatcherRunner.getDispatcherGateway();
+
+			testingLeaderElectionService.notLeader();
+			testingLeaderElectionService.isLeader(secondLeaderSessionId);
+
+			final CompletableFuture<DispatcherGateway> secondDispatcherGatewayFuture = dispatcherRunner.getDispatcherGateway();
+
+			assertThat(firstDispatcherGatewayFuture.get(), is(firstDispatcherGateway));
+			assertThat(secondDispatcherGatewayFuture.get(), is(secondDispatcherGateway));
+		}
+	}
+
+	@Test
+	public void getShutDownFuture_whileRunning_forwardsDispatcherLeaderProcessShutDownRequest() throws Exception {
+		final UUID leaderSessionId = UUID.randomUUID();
+		final CompletableFuture<ApplicationStatus> shutDownFuture = new CompletableFuture<>();
+		final TestingDispatcherLeaderProcess testingDispatcherLeaderProcess = TestingDispatcherLeaderProcess.newBuilder(leaderSessionId)
+			.setShutDownFuture(shutDownFuture)
+			.build();
+		testingDispatcherLeaderProcessFactory = TestingDispatcherLeaderProcessFactory.from(testingDispatcherLeaderProcess);
+
+		try (final DefaultDispatcherRunner dispatcherRunner = createDispatcherRunner()) {
+			testingLeaderElectionService.isLeader(leaderSessionId);
+
+			final CompletableFuture<ApplicationStatus> dispatcherShutDownFuture = dispatcherRunner.getShutDownFuture();
+
+			assertFalse(dispatcherShutDownFuture.isDone());
+
+			final ApplicationStatus finalApplicationStatus = ApplicationStatus.UNKNOWN;
+			shutDownFuture.complete(finalApplicationStatus);
+
+			assertThat(dispatcherShutDownFuture.get(), is(finalApplicationStatus));
+		}
+	}
+
+	@Test
+	public void getShutDownFuture_afterClose_ignoresDispatcherLeaderProcessShutDownRequest() throws Exception {
+		final UUID leaderSessionId = UUID.randomUUID();
+		final CompletableFuture<ApplicationStatus> shutDownFuture = new CompletableFuture<>();
+		final TestingDispatcherLeaderProcess testingDispatcherLeaderProcess = TestingDispatcherLeaderProcess.newBuilder(leaderSessionId)
+			.setShutDownFuture(shutDownFuture)
+			.build();
+		testingDispatcherLeaderProcessFactory = TestingDispatcherLeaderProcessFactory.from(testingDispatcherLeaderProcess);
+
+		try (final DefaultDispatcherRunner dispatcherRunner = createDispatcherRunner()) {
+			testingLeaderElectionService.isLeader(leaderSessionId);
+
+			final CompletableFuture<ApplicationStatus> dispatcherShutDownFuture = dispatcherRunner.getShutDownFuture();
+
+			assertFalse(dispatcherShutDownFuture.isDone());
+
+			dispatcherRunner.closeAsync();
+
+			final ApplicationStatus finalApplicationStatus = ApplicationStatus.UNKNOWN;
+			shutDownFuture.complete(finalApplicationStatus);
+
+			try {
+				dispatcherShutDownFuture.get(10L, TimeUnit.MILLISECONDS);
```

Comment:
The idea was to use a short timeout to allow for asynchronicity in the underlying implementation which we don't have atm. I think it can be simplified as you suggested.